Fixes #9412: Context command ignored implicit binders

herbelin opened the pull request:
**Kind:** bug fix 

Fixes / closes #9412

This is a quick fix (which will have to be adapted on top of #12272).

- [X] Added / updated test-suite
- [X] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

SkySkimmer commented on vernac/comAssumption.ml:
```diff
@@ -225,8 +233,10 @@ let context_nosection sigma ~poly ctx =
   in
   let fn subst d =
-    let (name,b,t,_impl) = context_subst subst d in
+    let (name,b,t,impl) = context_subst subst d in
```
Why this change? I don't see a use of `impl`.

herbelin replied:
It is an unwanted change which I forgot to revert. Thanks for noticing.

Nevertheless, it raises the question of what to do when `Context` occurs not in a section. Currently, it silently turns the declarations into `Axiom`/`Definition`. Should it, like `Variable`, raise a warning "Interpreting this declaration as if a global declaration prefixed by Local"? This is at least what the reference manual suggests in saying that `Context` "declare[s] variables in the context of the current section, like `Variable`". 

In any case, the `impl` should not be silently dropped when not in a section. It should be tested and a warning be raised if it is `Explicit` (i.e. if the user wrote `Context {x : T}` outside a section).

herbelin commented:
I added a warning for `Context` not in a section (but I'm not fully convinced; refining the refman to tell that `Context` outside a section is like the declarations made `Local` is maybe lighter).

Maybe the incompatibilities are too important and it'll be preferable to do the fix in two steps... A first release with a warning telling how to fix. The next release w/o the warning??

SkySkimmer commented:
Some people use context outside sections a lot, eg https://github.com/mit-pdos/perennial/blob/cdaa737e40d08c1c7c7d514db69a8f9b55814713/src/CSL/RefinementIdempotenceModule.v#L18-L52
Not sure we want to warn for that. Maybe an off-by-default warning?

herbelin commented:
> Some people use context outside sections a lot, eg https://github.com/mit-pdos/perennial/blob/cdaa737e40d08c1c7c7d514db69a8f9b55814713/src/CSL/RefinementIdempotenceModule.v#L18-L52
> Not sure we want to warn for that. Maybe an off-by-default warning?

Thanks. This confirms my view that we should rather change the doc. (But still keep a warning for `Context {a : A}` outside a section).

To deal with the compatibility, I'm now otherwise considering using the multiple sequence of implicit arguments feature, i.e. to do so that `Context (f : forall {x}, ...)` behaves as `Arguments f {x} ..., x ...` (including auto implicit arguments if any). This could even be generalized to _all_ manual declarations of implicit arguments, i.e., when the conclusion of `f` is rigid, that:
```
Definition f : forall {x}, ...
```
behaves (including auto implicit arguments) as:
```
Arguments f {x} ..., x ...
```
I suspect that this would be totally compatible, since the extra sequence would be used only when more arguments than expected are used.

Also, this would provide with an alternative way to explicitly give implicit arguments which is lighter than by using the `(x:=...)` notation.

What do you think?

herbelin commented:
> If this is meant to be a breaking change, the changelog entry should be much more explicit about this.

I started investigating how to make it  (hopefully) a non-breaking change, by using multiple lists of implicit arguments. It is not so clear how to do in an elegant way though:
- First, this has to exclude constants with flexible conclusion as e.g. in type `forall X, X -> X`, since this kind of constant have an arity difficult to control syntactically.
- Second, it would be too ad hoc to do it only for `Context`, so it would have to be done for all kinds of declarations.
- Third, it is not fully clear which range of flexibility we may want. For instance, in
  ```
  Set Implicit Arguments.
  Context (f : forall {A} B, A -> B -> Type).
  ```
  it would certainly be natural that:
  - with 4 arguments, it is `@f x y z t` (no implicit arguments at all)
  - with 2 arguments, it is `@f _ _ x y` (the automatic + manual implicit arguments)
  - with 1 argument, it is `@f _ _ x`
  - with 0 arguments, it is `@f _` (the maximal manual one)

  But with 3 arguments, should it fail on the principle that it is 2 implicit arguments or none? Or should it make a hierarchy between manual and automatic implicit arguments and consider either that the manual implicit arguments are stronger, giving `@f _ x y z`, or that the automatic implicit arguments are stronger, giving `@f x _ y z`. This is the point I arrived in my thinking.


[Slightly related, but for another PR, @mattam82 suggested me to deprecate automatic declaration of not-in-section `Context` assumptions as instances, as it was done already for `Axiom`.]

Zimmi48 commented:
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate, and also consider updating labels to reflect current status, moving the PR to draft status, etc.

We apologize in advance if we misunderstood the PR status.

herbelin commented:
This PR implements/fixes what I think is the expected behavior for `Context` and, to me, there is no doubt that it is good. The problem was to preserve compatibility. User and developer feedback welcome about how to manage the incompatibilities (a flag, an upgrading script, manual updates, ...)?

Zimmi48 commented:
I'm slightly worried about your ideas for making this a non-breaking change. I'd much rather advertise this as a breaking bug fix (like the one of `Export`) and implement the only correct behavior. Have you tried preparing overlays for the affected projects? Naively, I'd think that if the implicit binders were ignored, a backward-compatible fix is just to remove the implicit binders from these projects, wherever they create trouble. If backward-compatible overlays are easy to prepare, then the number of affected projects is not a problem: we should just make sure that the fix is properly advertised in the release notes. But there might be something I don't understand because I'm surprised that the compatibility impact is so large.

In any case: this seems slightly too late for 8.12, but this would still be great to finish this!

Zimmi48 commented:
BTW, the change to the doc is completely independent of the rest of the PR and should have been its own PR (it would have been merged much faster).

Zimmi48 commented:
The number of impacted projects seems so huge that I'm having a very hard time believing that there are really this many projects that used an implicit binder in the type of a variable declared in a `Context` command. It may instead be the effect of a buggy implementation...

SkySkimmer commented:
Why's the test suite broken?

herbelin commented:
> Why's the test suite broken?

Because I mistakenly pushed an experimental and non finalized commit living on top of my local copy of the branch. 😞 

The failures are otherwise not so many apparently: unimath, rewriter, cross-crypto, fiat-parsers.

jfehrle commented on doc/sphinx/language/core/sections.rst:
```diff
@@ -91,9 +91,14 @@ Sections create local contexts which can be shared across multiple definitions.
-   Declare variables in the context of the current section, like :cmd:`Variable`,
-   but also allowing implicit variables, :ref:`implicit-generalization`, and
-   let-binders.
+   When in a section, this combines several instances of
+   :cmd:`Variable` or :cmd:`Let` at once, also supporting
+   :ref:`implicit-generalization`. Variables which are surrounded by
+   %{ %} are turned into implicit arguments at the end of section.
+
+   When not in a section, this combines several instances of
+   :cmd:`Local Parameter` or :cmd:`Local Definition` at once, also
+   supporting :ref:`implicit-generalization`.
```
```suggestion
   Inside sections, combines several instances of
   :cmd:`Variable` or :cmd:`Let`, also supporting
   :ref:`implicit-generalization`. Variables which are surrounded by
   %{ %} are turned into implicit arguments at the end of section.

   Outside of section, combines several instances of
   :cmd:`Local Parameter` or :cmd:`Local Definition`, also
   supporting :ref:`implicit-generalization`.
```

But I have no idea what "combining" means and where the result of such combining would end up.  The "Declare variables" language previously used makes more sense (though would need to be adjusted to the inside/outside differences).

Is it also true that Context within a section is local to the section?

Zimmi48 replied:
I agree that the new description is more confusing than the previous one (which was mostly fine except that it didn't say what happened outside sections). But if we really want to highlight the use of `Context` outside sections, we should move this command out of the page on sections and next to the one of `Parameter` and `Variable`. To be discussed with the possible generalization of the `Parameter` and `Variable` syntax and also in relation to #13068. In any case, not really relevant to this PR.

>Is it also true that Context within a section is local to the section?

Yes.

Zimmi48 commented:
>The failures are otherwise not so many apparently: unimath, rewriter, cross-crypto, fiat-parsers.

That's a reasonable number then. The next step is fixing these to determine how easy / difficult it was without a warning.

herbelin commented:
Do we try to have this fix in 8.13? Note that overlays are not needed: the two upstream PRs remaining to be merged will maintain compatibility.

Zimmi48 commented:
Backward-compatible overlays are good because they avoid issues of synchronization and spurious failures on master. However, they don't change the fact that backporting a breaking change still requires updating the pinned commits on the release branch. So it really depends whether @gares thinks it's worth the effort or not.

BTW, how easy was it to fix the errors?

Zimmi48 commented on doc/changelog/02-specification-language/11390-master+fix-9412-context-implicits.rst:
```diff
@@ -0,0 +1,4 @@
+- **Fixed:**
+  Implicit arguments were ignored in :cmd:`Context`
```
Please add a `.. warning::` box that this has been observed to be a breaking change in several projects and explaining the various strategies to restore compatibility.

Zimmi48 reviewed (commented):
The changelog is not currently in a form that users will notice. It is important to highlight bug fixes that are breaking changes.

herbelin commented:
> BTW, how easy was it to fix the errors?

In most of cases, just removing the `{ }` was the intended things to do.

In other cases, the `{ }` were expected and this was simplifying the code. So as to be backward-compatible, I added extra (eventually superflous) `Local Arguments` declarations to compensate the non-declaration of implicit arguments, and I removed some `_` here and there.

In total, this was not a lot of changes. It took me about 2h though (doing other things in parallel) because the developments were long to compile. In total, it was 47 additions for fiat-crypto-legacy, 9 for rewriter, 2 for UniMath, 2 for fiat-parsers. I also fixed at the same time other failures of fiat-crypto-legacy in parts which the Coq CI was not checking, such as `Extraction Ocaml`/`Extraction OCaml` and `Hint Resolve` not generalizing anymore automatically over the unresolved evars.

Zimmi48 commented:
The changelog looks better. Note the comment of Jim on the documentation change.

herbelin commented:
> Note the comment of Jim on the documentation change.

I don't exactly what to do. Maybe is it useless to refer to `Variable` and `Let` (even more than `Variables` accept several declarations). Do we try to improve it now or will it change anyway later?

Zimmi48 commented:
Maybe revert the documentation change since it's not viewed as a clear improvement.

herbelin commented:
> Maybe revert the documentation change since it's not viewed as a clear improvement.

I agree. Actually, I'm not fuly convinced that `Context` outside a section is worth to be recommended, compared e.g. to `Parameters` and `Axioms`.

Zimmi48 commented:
> I agree. Actually, I'm not fuly convinced that `Context` outside a section is worth to be recommended, compared e.g. to `Parameters` and `Axioms`.

Before deciding whether to recommend it, I would like to clarify the status of #13068.

gares commented:
IMO this one goes to 8.14

Zimmi48 commented:
@gares: Segfault in the base+async job. See https://gitlab.com/coq/coq/-/jobs/866272033

Zimmi48 commented on doc/changelog/02-specification-language/11390-master+fix-9412-context-implicits.rst:
```diff
@@ -0,0 +1,15 @@
+- **Fixed:**
+  Implicit arguments were ignored in :cmd:`Context`
+  (`#11390 <https://github.com/coq/coq/pull/11390>`_,
+  by Hugo Herbelin, fixing `#9412 <https://github.com/coq/coq/pull/9412>`_).
+
+.. warning::
+
+   This has been observed to be a breaking change in a couple of
+   projects. This can be fixed either by removing the implicit
+   argument annotations, or by taking into account the new presence of
+   implicit arguments for the corresponding declarations. The first
+   approach provides backward compatibility. If needed, the second
+   approach can be made backward compatible too by adding an extra
+   (eventually superflous) :n:`Local Arguments` declaration right
+   after :cmd:`Context`.
```
This should fix the doc build issue.

```suggestion

  .. warning::

     This has been observed to be a breaking change in a couple of
     projects. This can be fixed either by removing the implicit
     argument annotations, or by taking into account the new presence of
     implicit arguments for the corresponding declarations. The first
     approach provides backward compatibility. If needed, the second
     approach can be made backward compatible too by adding an extra
     (eventually superflous) :n:`Local Arguments` declaration right
     after :cmd:`Context`.

  (`#11390 <https://github.com/coq/coq/pull/11390>`_,
  by Hugo Herbelin, fixing `#9412 <https://github.com/coq/coq/pull/9412>`_).
```

Zimmi48 commented:
fiat-crypto is failing but it looks like this is because you didn't include the overlay for this project in your overlay file.

gares commented:
> @gares: Segfault in the base+async job. See https://gitlab.com/coq/coq/-/jobs/866272033

> [ 4015.845895] coqc[13309]: segfault at 55a3e3f5a1cc ip 000055a3e3f5a1cc sp 00007ffe29464240 error 14
> [ 4015.860215] Code: Bad RIP value.

It's not an OOM, but a true segfault because it's fetching an instruction at a wrong address.
I've no idea why/how this can happen.

jfehrle commented:
> I've no idea why/how this can happen.

If rerunning the job clears the segfault, the error was probably caused by cosmic rays flipping bits in DRAM.  Some high-end servers have error correcting memory to address this.  They could fix single-bit errors and detect 2-bit errors in a single word of memory.  IIRC this required 5 extra bits for every 32 (16??) bits of memory.  Another alternative was to have a parity bit that could detect a single-bit bit error.  Both of these require additional hardware.

The theory and practice of error correction is interesting and very sophisticated.  It's essential to disk drives and communications. This [article](https://en.wikipedia.org/wiki/ECC_memory) mentions that the Cassini-Huygens deep space probe corrected about 280 single-bit errors per day.  It also mentions a Google study that showed that "More than 8% of DIMM memory modules were affected by errors per year" in its servers.

gares commented:
Thanks for the support. Here it is believed that there is a correlation between "async" and these failures.

Async uses threads (in a very trivial ways), doubles the number of processes (hence memory usage), and marshals data between processes (the only unsafe feature of OCaml we use).
I'd love cosmic rays to be the culprit, but I'm afraid it's too often reproducible to be that. Even if I've not clue what the reason could be :-/

The async job tests what happens when you open CoqIDE (or an other XML-protocol based UI) and jump to the end of the file. This file on the Relas likes to crash.

Soon (as in months) we will have a new document manager which does not use threads (but still forks processes and marshals data). If we sill still see this things, then I'll but 100% your thesis ;-)

Zimmi48 commented:
Note that the suggestion to revert the changes to the documentation is still not addressed.

herbelin commented:
> Note that the suggestion to revert the changes to the documentation is still not addressed.

I was waiting for more feedback... 

But we can also take the decision only the two of us...

Zimmi48 commented:
Feedback on the wording or feedback on whether to advertise `Context` outside sections or not?

herbelin commented:
Feedback on going in the direction of not documenting `Context` outside sections.

Zimmi48 commented:
> Feedback on going in the direction of not documenting `Context` outside sections.

OK, then it's a discussion that is worth having with more than two developers IMHO but definitely *not* in this orthogonal PR.

And if we do decide to advertise `Context` beyond sections, then it's more than an additional paragraph that we need: it's a change of location of the `Context` documentation.

Let's take this to the Coq Call.

herbelin commented:
I then removed the doc commit so as to split the different issues.

Zimmi48 commented:
The version you pushed doesn't build anymore.

herbelin commented:
> The version you pushed doesn't build anymore.

What a relief! I was looking everywhere where my ongoing work on #13468 had gone and it was just accidentally committed to this branch!

Alizter commented:
ping @herbelin what is to be done with this?

Zimmi48 commented:
@herbelin Was this PR almost ready for the last two years?
